sympy/matrices/normalforms.py

```diff
@@ -0,0 +1,136 @@
+    Return the tuple of abelian invariants for a matrix `m`
+    (as in the Smith-Normal form)
+
+    '''
+    if not domain:
+        if not (hasattr(m, "ring") and m.ring.is_PID):
+            raise ValueError(
+                "The matrix entries must be over a principal ideal domain")
+        else:
+            domain = m.ring
+
+    m = m[:, :]
+
+    def add_rows(m, i, j, a, b, c, d):
+        # replace m[i, :] by a*m[i, :] + b*m[j, :]
+        # and m[j, :] by c*m[i, :] + d*m[j, :]
+        for k in range(m.cols):
+            e = m[i, k]
+            m[i, k] = a*e + b*m[j, k]
+            m[j, k] = c*e + d*m[j, k]
+
+    def add_columns(m, i, j, a, b, c, d):
+        # replace m[:, i] by a*m[:, i] + b*m[:, j]
+        # and m[:, j] by c*m[:, i] + d*m[:, j]
+        for k in range(m.rows):
+            e = m[k, i]
+            m[k, i] = a*e + b*m[k, j]
+            m[k, j] = c*e + d*m[k, j]
+
+    def clear_column(m):
+        # make m[1:, 0] zero by row and column operations
+        if m[0,0] == 0:
+            return m
+        pivot = m[0, 0]
+        for j in range(1, m.rows):
+            if m[j, 0] == 0:
+                continue
+            d, r = domain.div(m[j,0], pivot)
+            if r == 0:
+                add_rows(m, 0, j, 1, 0, -d, 1)
+            else:
+                a, b, g = domain.gcdex(pivot, m[j,0])
+                d_0 = domain.div(m[j, 0], g)[0]
+                d_j = domain.div(pivot, g)[0]
+                add_rows(m, 0, j, a, b, d_0, -d_j)
+                pivot = g
+        return m
+
+    def clear_row(m):
+        # make m[0, 1:] zero by row and column operations
+        if m[0] == 0:
+            return m
+        pivot = m[0, 0]
+        for j in range(1, m.cols):
+            if m[0, j] == 0:
+                continue
+            d, r = domain.div(m[0, j], pivot)
+            if r == 0:
+                add_columns(m, 0, j, 1, 0, -d, 1)
+            else:
+                a, b, g = domain.gcdex(pivot, m[0, j])
+                d_0 = domain.div(m[0, j], g)[0]
+                d_j = domain.div(pivot, g)[0]
+                add_columns(m, 0, j, a, b, d_0, -d_j)
+                pivot = g
+        return m
+
+    ind = [i for i in range(m.rows) if m[i,0] != 0]
+    if ind:
+        m = m.permute_rows([[0, ind[0]]])
+    else:
+        ind = [j for j in range(m.cols) if m[0,j] != 0]
+        if ind:
+            m = m.permute_cols([[0, ind[0]]])
+
+    while any([m[0,i] != 0 for i in range(1,m.cols)]) or any([m[i,0] != 0 for i in range(1,m.rows)]):
+        m = clear_column(m)
+        m = clear_row(m)
+
+    if 1 in m.shape:
+        invs = []
+    else:
+        invs = smith_normal_invariants(m[1:,1:], domain=domain)
+    zeros = []
+    result = [m[0]]
+    for r in invs:
+        if r == 0:
+            zeros.append(r)
+        else:
+            result.append(r)
+    # in case m[0] doesn't divide the invariants of the rest of the matrix
+    if result[0]:
+        for i in range(len(result)-1):
```

Comment:
Thanks for the explanation, that is about what I was expecting. It seems that it is valid though I don't recall having seen that algorithm explained anywhere. Can you find a reference? There should be references for algorithms used in SymPy, especially for those that are not as familiar as Euclid's algorithm. Othewise there should be adequate comments along the code, which at this point might become rather longish.

I'd also suggest that that the zero invariants would come last. (That would be natural as I noted above.) Then it seems that it would not be necessary to scan `inv` separating zeros and nonzero results. It would suffice to run this loop as long as the invariants are nonzero before breaking (if `inv[i + 1] == 0`, then it is divisible by `inv[i]`).